feat: MultiEnv step added

tapaszto opened the pull request:
MultiEnv step is new workflow step, a combination of RunStep and EnvStep, executes a command returning an array of name-value pairs, in case of success these pairs are getting added as environment variables before running plan and apply otherwise the errorMessage is getting displayed in the output.
It allows adding dynamic number of envionment variables depending on the command execution result.

The result of executed command must have a fixed format:
`{
  "success":true,
  "errorMessage":null,
  "result":
    [
      {"name":"TF_VAR_VARIABLE_ONE","value":"value1"},
      {"name":"TF_VAR_VARIABLE_TWO","value":"value2"},
      {"name":"TF_VAR_VARIABLE_THREE","value":"value3"}
    ]
}`

Using in workflow definition:
`- multienv: bash -c /home/tapaist/atlantisbuild/config/multienv.sh`

Our use case that requires this new feature:
Atlantis is being used by several teams within our company. The teams are not allowed to override the global workflow defined in the server config as it contains specific checks and other special items that are required for our Atlantis instance to run e.g. defining the required Azure credentials which is always tied to the selected repository-Azure subscription. The various teams often have the requirement to be able to define additional environment variables, mostly secrets that need to be used for Terraform apply. These secrets are team dependent and must be isolated, one team must not be able to use the secrets of an other team. Overriding the workflow in repo config to have custom team dependent env steps is not permitted and we cannot have dozens of team/repo dependent workflows on server side with continuously increasing count. We need to have one single point in the workflow config on server side where the dynamic number of team/repo specific environment variables with dynamic name/value pairs can be injected safely from client side. However we've got the env step but it can only define one single environment variable, and also the name of it is hardwired in the config.

The solution:
Due to the new multienv step multiple environment variables can be added dynamically just before executing Terraform commands via Atlantis. We specifically use this step in Atlantis workflow to call a Linux script hosted on Atlantis server, the script reads the reference to an Azure key vault from a file with specific name KeyVault.ref committed to the repo and having authorization reads the secrets from the key vault. In our case the Linux script is nothing more than a bridge to call further functionality and the committed extra file in the repo does not contain any sensitive info but a name of a key vault. All the sensitive info comes from a safe key vault on-the-fly and the result is passed to Atlantis workflow on server side as environment variables similarly to the original env workflow step. By this method the different teams can define one single KeyVault.ref file per repo, they enlist the required environment variable name-value pairs in the key vault, grant the repo dependent Azure service principal the required permission to read the key vault and Atlantis injects the secrets (as environment variables) that can be referenced in the Terraform scripts.

tapaszto commented:
Hi @acastle,

Can You review this please?

jamengual commented:
Thanks @tapaszto for your contribution!

few comments:
New features need documentation
Tests are required too

and can you please expand on the problem this functionality is trying to solve?

Thanks.

tapaszto commented:
Hi @jamengual / @nishkrishnan,

I added the requested automated test and updated the documentation and also expanded the PR with the details of our use case. Please check my updated opening comment on the top.

tapaszto commented:
> Sorry for getting to this super late, my overall thoughts on this:
> 
> Instead of having another run step (multienv) can we just make a breaking change to the env step so it'd look like:
> 
> ```yaml
> env:
>   # this is the command you're implementing and by having under run, we can also have a file option as well.
>   run: some_command.sh
>   # legacy implementation
>   vars:
>     - name: <name>
>        value: <value>
> ```
> 
> We would then basically just release a new major version of Atlantis at this point.
> 
> Another thing I wanted to bring out was the format of the resulting env vars. I think in general having json be a requirement of the run command can be annoying for users that just want to write simple scripts. Would like us to just simplify here and have a simple multi-line env key pair output.. I don't think we need other fields such as success and error message as we'll just base errors off non-zero exit codes like the rest of the run commands in atlantis.
> 
> so the output of the command could look like:
> 
> ```
> ENV_VAR1=VALUE1
> ENV_VAR2=VALUE2
> ```
> 
> where name and value are separated by "=" (standard env var definition) and key pairs are separated by new lines.
> 
> What do you think?

We are concerned that your proposal for modifying the existing env step with the breaking change is too complex for us and considering the limited resources we have, we are advising a simpler implementation which fulfils the requirement to avoid creating a new run step (multienv step) but does not require any breaking change in the implementation: the original env step has got name, value and command attributes, let the env step be as it is now + in case of the name and value equal to a reserved constant e.g. name == "dynamic" && value == "dynamic" the step would  call the command attribute and add the resulted variable number of environment variables to the workflow and  would work as we previously implemented the multienv step. By this we can fulfil your request to avoid having a new run step but we would not have a breaking chance. We are capable to do either this or our original implementation with the new multienv step. What is your view regarding our suggestion?

Dilergore commented:
Hi @nishkrishnan,

Could you please take a look at the proposal above? We need this change in our environment.

Thanks!

tapaszto commented:
Hi @nishkrishnan , @lkysow, @acastle ,

Guys, can you please answer the proposal above? We really need this feature.

tapaszto commented:
Hi,

Can you check the proposal above please?

jamengual commented:
@tapaszto Hi, we talked about your feature and usually we try to merge features that are popular and requested by people and it seems that you guys are the only ones doing this way, maybe? if you could get some votes and maybe other people to comment we could potentially merge this feature. Thanks.

Dilergore commented:
@jamengual There was a feature request earlier here - Multiple people for seeking for this:
https://github.com/runatlantis/atlantis/issues/1769

Theoretically I could start to ask our entire firm to start to comment here I do not know what change that will make. We are trying to add here a feature which could be used in many ways and would make Atlantis more flexible.

TBH We contributed already and planned to contribute more but this is really not encouraging us to do so. We do similar activities for other open source projects and the way I see Atlantis is the most problematic and the slowest one.

atakacs90 commented:
Would like to use this feature as well.

szferle commented:
It would be useful for us as well

hd214 commented:
I need it to.

apoorvabelsare commented:
This feature will be very useful for us.

jamengual commented:
> @jamengual There was a feature request earlier here - Multiple people are seeking this funtionality: #1769
> 
> Theoretically I could start to ask our entire firm to start to comment here I do not know what change that will make. We are trying to add here a feature which could be used in many ways and would make Atlantis more flexible.
> 
> TBH We contributed already and planned to contribute more but this is really not encouraging us to do so. We do similar activities for other open source projects and the way I see Atlantis is the most problematic and the slowest one.

I'm sorry that you feel not encouraged to contribute, we are just 5 people with full-time jobs trying to help the project to the best of our abilities so time is very restrictive right now.

As you might know from your experience with other projects there is always a popularity aspect to the new features so that is why I ask about the 👍🏽 , how those happen is not important.

There are a few comments on this PR that need to be addressed still so I hope once those are resolved we can have another review and see if we can move forward.

Thanks.
@tapaszto

jamengual commented:
@tapaszto please resolve the conflict when you have some time, thanks.

tapaszto commented:
Hi @jamengual,

The multienv_step_runner_test.go file and the test logic are fixed, and the conflict is resolved.
Can you merge the PR please?

jamengual commented:
@tapaszto I can't do that, the review process is a team effort, I will bring this up with the team again for a review, please allow for a few weeks before we get back to you.

jamengual commented:
@tapaszto please reply to the comments in code or resolve the conversations.

tapaszto commented on server/core/runtime/multienv_step_runner_test_go:
```diff
@@ -0,0 +1,77 @@
+package runtime_test
```
Fixed

tapaszto replied:
fixed

tapaszto commented:
> Thanks @tapaszto for your contribution!
> 
> few comments: New features need documentation Tests are required too
> 
> and can you please expand on the problem this functionality is trying to solve?
> 
> Thanks.

Documentation and test are added & more detailed problem explanation is included

tapaszto commented on server/core/runtime/multienv_step_runner.go:
```diff
@@ -0,0 +1,53 @@
+					sb.WriteString("Dynamic environment variables added:\n")
+					for _, item := range callerResult.Result {
+						envs[item.Name] = item.Value
+						sb.WriteString(item.Name)
```
Please let me share why we need so detailed logging.
Within our company we have a dedicated team that technically owns Atlantis (and other stuffs, but this is not relevant), only this team has the knowledge how Atlantis works, how the architecture looks like. There are dozens of other teams whose Terraform developer members only know that commenting to their Azure DevOps PRs plans and applies Terraform related infra. They only aware of their own repo and a connected Azure subscription. An automated self service ticketing mechanism is provided for them to select a repo and assign to a Azure subscription (with proper permission handling), so they does not even know how their own repo is connected to their subscription, they only focus on writing Terraform code in order to set up their own Azure infra.
Let's see our current use case. Somebody needs a team specific secret in his/her Terraform code e.g. a specific Terraform provider requires it in the form of an environment variable (this is a very common scenario). We (the Atlantis owner hub team) do not want to add project specific envvars to our company specific Atlantis workflow (containing custom checks by 3rd party tools integrated), and the members of the specific project does not have the permission to create their own workflow (to avoid overriding the company standards/policy).
Instead, this the team member can pass the secret via a specific file in the repo that references a key vault hosting the secrets. He does not know anything but he needs to have a committed file in his repo, the file contains the name of a key vault and the access permissions need to be set on the key vault properly. Atlantis workflow extracts the secrets from the key vault via a web api call and creates the required environment variables under the hood.
The webservice must perform several actions where multiple errors can occur, e.g.:
- The committed key vault file contains invalid reference
- The key vault does not exist
- The team specific Azure deployment user does not have the proper permission to access the key vault
- Missing permission to read the secrets from the key vault
- Atlantis workflow is not able to call the web api, maybe network issue
These notifications must be passed to the UI (which is the Azure DevOps PR in our case) to let the users resolve the issues.
Sometimes the key vault reference is correct and the permissions to access the key vault are also correct, Atlantis workflow successfully creates the environment variables, but Terraform does not find the specific envvar. Maybe the user unintentionally referenced an other key vault or meanwhile somebody else removed the required secrets from the vault.
That is why it's essential to display the dynamic envvar names through Atlantis PR comment, and let the user view what kind of dynamic envvars are provided to Terraform.
Anyway, according to our security review displaying the name (only the name, not he value!) of the dynamic envvars is not a big security risk compared what we can win, our goal is providing a self service Azure infra building process, avoiding the tickets containing "why the hell Terraform cannot see the envvars" that probaly would lead to manual intervention one by one wasting our resources.

tapaszto commented on server/core/runtime/multienv_step_runner.go:
```diff
@@ -0,0 +1,53 @@
+					}
+					return sb.String(), nil
+				}
+				return "No dynamic environment variable added", nil
```
Please see my comment below.

tapaszto commented:
> > Sorry for getting to this super late, my overall thoughts on this:
> > Instead of having another run step (multienv) can we just make a breaking change to the env step so it'd look like:
> > ```yaml
> > env:
> >   # this is the command you're implementing and by having under run, we can also have a file option as well.
> >   run: some_command.sh
> >   # legacy implementation
> >   vars:
> >     - name: <name>
> >        value: <value>
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > We would then basically just release a new major version of Atlantis at this point.
> > Another thing I wanted to bring out was the format of the resulting env vars. I think in general having json be a requirement of the run command can be annoying for users that just want to write simple scripts. Would like us to just simplify here and have a simple multi-line env key pair output.. I don't think we need other fields such as success and error message as we'll just base errors off non-zero exit codes like the rest of the run commands in atlantis.
> > so the output of the command could look like:
> > ```
> > ENV_VAR1=VALUE1
> > ENV_VAR2=VALUE2
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > where name and value are separated by "=" (standard env var definition) and key pairs are separated by new lines.
> > What do you think?
> 
> We are concerned that your proposal for modifying the existing env step with the breaking change is too complex for us and considering the limited resources we have, we are advising a simpler implementation which fulfils the requirement to avoid creating a new run step (multienv step) but does not require any breaking change in the implementation: the original env step has got name, value and command attributes, let the env step be as it is now + in case of the name and value equal to a reserved constant e.g. name == "dynamic" && value == "dynamic" the step would call the command attribute and add the resulted variable number of environment variables to the workflow and would work as we previously implemented the multienv step. By this we can fulfil your request to avoid having a new run step but we would not have a breaking chance. We are capable to do either this or our original implementation with the new multienv step. What is your view regarding our suggestion?

tapaszto commented:
> > > Sorry for getting to this super late, my overall thoughts on this:
> > > Instead of having another run step (multienv) can we just make a breaking change to the env step so it'd look like:
> > > ```yaml
> > > env:
> > >   # this is the command you're implementing and by having under run, we can also have a file option as well.
> > >   run: some_command.sh
> > >   # legacy implementation
> > >   vars:
> > >     - name: <name>
> > >        value: <value>
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > We would then basically just release a new major version of Atlantis at this point.
> > > Another thing I wanted to bring out was the format of the resulting env vars. I think in general having json be a requirement of the run command can be annoying for users that just want to write simple scripts. Would like us to just simplify here and have a simple multi-line env key pair output.. I don't think we need other fields such as success and error message as we'll just base errors off non-zero exit codes like the rest of the run commands in atlantis.
> > > so the output of the command could look like:
> > > ```
> > > ENV_VAR1=VALUE1
> > > ENV_VAR2=VALUE2
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > where name and value are separated by "=" (standard env var definition) and key pairs are separated by new lines.
> > > What do you think?
> > 
> > 
> > We are concerned that your proposal for modifying the existing env step with the breaking change is too complex for us and considering the limited resources we have, we are advising a simpler implementation which fulfils the requirement to avoid creating a new run step (multienv step) but does not require any breaking change in the implementation: the original env step has got name, value and command attributes, let the env step be as it is now + in case of the name and value equal to a reserved constant e.g. name == "dynamic" && value == "dynamic" the step would call the command attribute and add the resulted variable number of environment variables to the workflow and would work as we previously implemented the multienv step. By this we can fulfil your request to avoid having a new run step but we would not have a breaking chance. We are capable to do either this or our original implementation with the new multienv step. What is your view regarding our suggestion?

Considering that 3 months passed since we created the PR we feel that creating a new implementation would not be effective, this would restart the whole discussion and review process. Please check our original implementation. This is a small but important feature for us as it's starting to be a showstopper in the case of specific projects and the cloud related infra deployment.

tapaszto commented:
Please let me share why we need so detailed logging.
Within our company we have a dedicated team that is technically owns Atlantis (and other stuffs, but this is not relevant), only this team has the knowledge of the how Atlantis works, how the architecture looks like, and there are dozens of other team whose terraform developer members only know that commenting to their Azure DevOps PRs plans and applies terraform related infra, they only aware of their own repo and a connected Azure subscription. An automated self service ticketing mechanism is provided for them to select a repo and assign to a Azure subscription (with proper permission handling), so they does not even know how their own repo is connected to their subscription, they only focus on writing Terrafom code in order to set up their on Azure infra.
Let's see our current use case. Somebody needs a team specific secret in his/her Terraform code e.g. a specific Terraform provider requires it in the form of environment variable (this is a very common scenario). We (the Atlantis owner hub team) do not want to add project specific envvars to our company specific Atlantis workflow (containing custom checks with 3rd party tools integrated to Atlantis workflow), and the members of the specific project does not have the permission to create their own workflow (to avoid overriding the company standards/policy).
Instead of this the team member can pass the secret via a specific file in the repo that references a KeyVault hosting the secrets. He does not know anything just that he needs to have a committed file in his repo, the file contains the name of a KeyVault and the access permissions need to be set on the KeyVault. Atlantis workflow extracts the secrets from the keyvault via a web api call and creates the required environment variables under the hood.
The webservice must perform several actions where multiple errors can occur, e.g.:
- The committed keyvault file contains invalid reference
- The keyvault does not exist
- The team specific Azure deployment user does not have the proper permission to access the keyvault
- Missing permission to read the secrets from the keyvault
- Atlantis workflow is not able to call the web api, maybe network issue
These notifications must be bassed to the UI (which is the Azure DevOps PR in our case) to let the users resolve the issues.
Sometimes the keyvault reference is correct and the permissions to access the keyvault are also correct, Atlantis worklow successfully creates the environment variables, but Terraform does not find a psecific envvar. Maybe the user unintentionally referenced an other keyvault or meanwhile somebody else removed the required secrets from the vault.
That is why it's essential to display the envvar name through Atlantis PR comment, and let the user view what kind of dynamic envvars hi provided to Terraform.

peterkulik commented:
> Please let me share why we need so detailed logging. Within our company we have a dedicated team that technically owns Atlantis (and other stuffs, but this is not relevant), only this team has the knowledge of how Atlantis works, how the architecture looks like, and there are dozens of other team whose terraform developer members only know that commenting to their Azure DevOps PRs plans and applies terraform related infra, they only aware of their own repo and a connected Azure subscription. An automated self service ticketing mechanism is provided for them to select a repo and assign to an Azure subscription (with proper permission handling), so they does not even know how their own repo is connected to their subscription, they only focus on writing Terraform code in order to set up their on Azure infra. Let's see our current use case. Somebody needs a team specific secret in his/her Terraform code e.g. a specific Terraform provider requires it in the form of environment variable (this is a very common scenario). We (the Atlantis owner hub team) do not want to add project specific envvars to our company specific Atlantis workflow (containing custom checks with 3rd party tools integrated to Atlantis workflow), and the members of the specific project do not have the permission to create their own workflow (to avoid overriding the company standards/policy). Instead, the team member can pass the secret via a specific file in the repo that references a key vault hosting the secrets. He does not know anything just that he needs to have a committed file in his repo, the file contains the name of a key vault and the access permissions needs to be set on the key vault. Atlantis workflow extracts the secrets from the key vault via a web api call and creates the required environment variables under the hood. The webservice must perform several actions where multiple errors can occur, e.g.:
> 
> * The committed key vault file contains invalid reference
> * The key vault does not exist
> * The team specific Azure deployment user does not have the proper permission to access the key vault
> * Missing permission to read the secrets from the key vault
> * Atlantis workflow is not able to call the web api, maybe network issue
> 
> These notifications must be passed to the UI (which is the Azure DevOps PR in our case) to let the users resolve the issues. Sometimes the key vault reference is correct and the permissions to access the key vault are also correct, Atlantis workflow successfully creates the environment variables, but Terraform does not find a specific envvar. Maybe the user unintentionally referenced an other key vault or meanwhile somebody else removed the required secrets from the vault. That is why it's essential to display the envvar name through Atlantis PR comment, and let the user view what kind of dynamic envvars hi provided to Terraform.

Hi @tapaszto, according to your use-case above you call a webservice by your new workflow step and pass secrets from client side to Atlantis workflow dynamically. It would be a useful feature for us, but I cannot see how you can call a 3rd party webservice by this multienv step, can you describe the details please?

Dilergore commented:
Hi @peterkulik,

We have some curl commands as part of bash scripts. We call the script as a run steps to return the variable value(s).

tapaszto commented:
> Hi @peterkulik,
> 
> We have some curl commands as part of bash scripts. We call the script as a run step to return the variable value(s).

Some more details.
We added a multienv step to our custom Atlantis workflow which executes the multienv.sh script, this is called in any case. Here is an example of the customized server side workflow:

```
workflows:
  default:
    plan:
      steps:
      - init
      - multienv: bash -c /home/tapaist/atlantisbuild/config/multienv.sh
      - plan
    apply:
      steps:
      - multienv: bash -c /home/tapaist/atlantisbuild/config/multienv.sh
      - apply
```

The execution of the multienv.sh check whether there is a specific file (called KeyVault.ref) in the repo, if yes then calls a specific webservice, gets the result and echoes back to the multienv step which creates the resulted environment variables dynamically or adds the error message the pull request output.

Here is an example json what the multienv workflow step expects from the webservice:

```
{
  "Success": true,
  "ErrorMessage": "", 
  "Result":
    [
      { "name":"TF_VAR_REPODEFINEDVARIABLE_ONE","value":"value1"}
    ]
}
```

peterkulik commented:
> > Hi @peterkulik,
> > We have some curl commands as part of bash scripts. We call the script as a run step to return the variable value(s).
> 
> Some more details. We added a multienv step to our custom Atlantis workflow which executes the multienv.sh script, this is called in any case. Here is an example of the customized server side workflow:
> 
> ```
> workflows:
>   default:
>     plan:
>       steps:
>       - init
>       - multienv: bash -c /home/tapaist/atlantisbuild/config/multienv.sh
>       - plan
>     apply:
>       steps:
>       - multienv: bash -c /home/tapaist/atlantisbuild/config/multienv.sh
>       - apply
> ```
> 
> The execution of the multienv.sh check whether there is a specific file (called KeyVault.ref) in the repo, if yes then calls a specific webservice, gets the result and echoes back to the multienv step which creates the resulted environment variables dynamically or adds the error message to the pull request output.
> 
> Here is an example json what the multienv workflow step expects from the webservice:
> 
> ```
> {
>   "Success": true,
>   "ErrorMessage": "", 
>   "Result":
>     [
>       { "name":"TF_VAR_REPODEFINEDVARIABLE_ONE","value":"value1"}
>     ]
> }
> ```

Nice one!
This suits for our case as we would implement the additional logic in the webservice

Ledermayer commented:
This sounds like a pretty great feature, especially where there are a lots of different teams are using atlantis.
Looking forward to have this merged.

tapaszto commented:
> I guess this new workflow step does not contain any Azure specific parts and can be used with non-Azure repos as well, am I right?

Yes, it's a fully general step. Please check the Atlantis supported VCs [here](https://www.runatlantis.io/docs/requirements.html#git-host)

pranavdesai1900 commented:
Would really like to have this feature.

PadreGregor commented:
really useful , looking forward to it

ko-be commented:
We've tested this out, rebased against the current latest release (v0.18.2) and it appears to be working well.  It does appear to have conflicts against the current master branch, however.

This feature would also be extremely helpful for us and is currently blocking us from fully implementing Atlantis in our environment as well.

Dilergore commented:
Hi @jamengual,

Have you had a chance to get this reviewed with your teammates?

Now that @tapaszto solved the conflicts you could merge this in.

Thanks!

Dilergore commented:
Hi @jamengual,

Have you had the meeting with the team?

jamengual commented:
Yes sir.

In about a week I will have more details but I think is looking good, one thing we did talk about was the output format, I think it will be better if the output is like this:

```
KEY=val
KEY2=val2
```​
instead of being opinionated and using json, it is simple to JSON encode the output if the user needs it.

Thanks.

jamengual commented:
@tapaszto do you think you can change the output to key value of ENV variables instead of a json output?

then we can merge this.

Thanks.

tapaszto commented:
> * The committed key vault file contains invalid reference
> * The key vault does not exist
> * The team specific Azure deployment user does not have the proper permission to access the key vault
> * Missing permission to read the secrets from the key vault
> * Atlantis workflow is not able to call the web api, maybe network issue

The error message is also very important for us as the called web service performs several checks and passing the error to the end users is essential, I added a very detailed comment 2 moths ago, just to summarize we need to handled cases such as:

1. The committed key vault file contains invalid reference
2. The key vault does not exist
3. The team specific Azure deployment user does not have the proper permission to access the key vault
4. Missing permission to read the secrets from the key vault
5. Atlantis workflow is not able to call the web api, maybe network issue

Can we add a special ErrorMessage key as well for error handling purposes? This would not be a valid envvar.

jamengual commented:
Hi @tapaszto 

I think that should be added to another PR to be discussed, I definitely do not recommend adding functionality to this PR otherwise we will need to review it again and it is just never going to get merged.

If you can address the comment I made earlier about the json output then we can work on merging this and you can open a PR for the error handling after.

Thanks.

jamengual commented:
@tapaszto we will like to merge and release this on the `prerelease` image we are preparing, do you think you will have time to make those changes?

jamengual commented:
Does anyone else want to take this and make the changes?

tapaszto commented:
Hi @jamengual,

Sorry, I was assigned to other projects I will complete this

tapaszto commented:
Hi @jamengual,

I performed the requested modifications. The website_link_check fails but I think this is not related to my changes.
Can you merge the PR please?

jamengual commented on runatlantis.io/docs/custom-workflows.md:
```diff
@@ -407,3 +407,24 @@ as the environment variable value.
+The result of the executed command must have a fixed format:
+EnvVar1Name=value1,EnvVar2Name=value2,EnvVar3Name=value3
+
+The name-value pairs in the result are getting added as envinronment variables if success is true otherwise the workflow execution stops with error and the errorMessage is getting displayed.
```
```suggestion
The name-value pairs in the result are added as environment variables if success is true otherwise the workflow execution stops with error and the errorMessage is getting displayed.
```

jamengual commented:
> 

We will do one last review and we will then merge it to be released on the `pre-release` then we will leave the `pre-release` available for 2-4 weeks and if we do not get any issues reported we will release it in the base release.

jamengual commented:
https://github.com/runatlantis/atlantis/releases/tag/v0.19.4-pre.20220513
